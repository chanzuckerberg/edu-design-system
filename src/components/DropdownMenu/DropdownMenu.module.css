@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    #DROPDOWN
\*------------------------------------*/

/**
 * Define themable properties
 */

/**
 * 1) Dropdown menu
 * 2) Set a minimum width for the dropdown menu
 */
.dropdown-menu {
  width: 16rem; /* 2 */
  background: var(--eds-theme-color-background-neutral-default);
  border: 1px solid var(--eds-theme-color-border-neutral-subtle);
  border-radius: var(--eds-theme-border-radius);
  box-shadow: var(--eds-theme-box-shadow);

  /**
  * Dropdown menu within active dropdown menu item
  */
  .dropdown-menu__item.eds-is-active & {
    visibility: visible;
    transition-delay: 0s;
  }

  /**
  * Dropdown menu within dropdown menu item align top left
  * 1) Position nested dropdown menu to the left and aligned at the bottom of dropdown menu item
  */
  .dropdown-menu__item--align-top-left > & {
    top: auto;
    bottom: 0;
  }
}

/**
 * Dropdown menu list
 */
.dropdown-menu__list {
  @mixin reset;
  list-style: none;
}

/**
 * Dropdown menu item
 */
.dropdown-menu__item {
  position: relative;
  width: 100%;
}

/**
 * Dropdown menu link
 */
.dropdown-menu__link {
  @mixin eds-theme-typography-body-text-sm;
  display: grid;
  grid-template-columns: auto 1fr;
  gap: var(--eds-size-1);
  appearance: none;
  background: none;
  text-decoration: none;
  border: none;
  width: 100%;
  padding: var(--eds-size-1-and-half) var(--eds-size-2);
  text-align: left;

  color: var(--eds-theme-color-text-neutral-default);
  transition: all var(--eds-anim-fade-quick) var(--eds-anim-ease);

  &:hover,
  &:focus {
    background: var(--eds-theme-color-background-neutral-subtle);
  }

  &:focus-visible {
    @mixin focus;
  }

  /**
  * Dropdown menu link within a lined dropdown menu item
  * 1) Add bottom border to the item
  */
  .dropdown-menu__item--lined & {
    border-bottom: 1px solid var(--eds-theme-color-border-neutral-subtle);
  }

  .dropdown-menu__item--error & {
    color: var(--eds-theme-color-text-utility-error);
  }
}
