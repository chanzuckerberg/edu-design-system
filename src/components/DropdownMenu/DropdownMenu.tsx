import clsx from 'clsx';
import React, {
  ReactNode,
  createContext,
  useRef,
  useEffect,
  KeyboardEvent,
  HTMLAttributes,
} from 'react';
import styles from './DropdownMenu.module.css';
import {
  L_ARROW_KEYCODE,
  U_ARROW_KEYCODE,
  R_ARROW_KEYCODE,
  D_ARROW_KEYCODE,
  ESCAPE_KEYCODE,
} from '../../util/keycodes';

export type Props = {
  /**
   * Child node(s) that can be nested inside component
   */
  children?: ReactNode;
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Sets the component to open or close by default
   */
  isActive?: boolean;
  handleOnKeyDown?: (e: React.KeyboardEvent) => void;
} & HTMLAttributes<HTMLElement>;

export const DropdownMenuContext = createContext<any>({});
/**
 * BETA: This component is still a work in progress and is subject to change.
 *
 * ```ts
 * import {DropdownMenu} from "@chanzuckerberg/eds";
 * ```
 *
 * Dropdown menu with actions list
 */
export const DropdownMenu: React.FC<Props> = ({
  children,
  className,
  isActive,
  handleOnKeyDown,
  ...other
}) => {
  type Refs = {
    set: Set<HTMLLIElement>;
    list: HTMLLIElement[];
  };
  const refs = useRef<Refs>({
    set: new Set(),
    list: [],
  });

  let focusIndex = 0;
  const focusItem = (index: number) => {
    refs.current.list[index]
      ?.querySelector<HTMLButtonElement | HTMLAnchorElement>(':first-child')
      ?.focus();
  };

  useEffect(() => {
    if (isActive && refs.current.list.length) {
      focusItem(0);
    }
  }, [isActive]);

  const onKeyDown = (e: KeyboardEvent<HTMLUListElement>) => {
    if ([ESCAPE_KEYCODE].includes(e.key)) {
      if (handleOnKeyDown) {
        handleOnKeyDown(e);
      }
    }

    if ([R_ARROW_KEYCODE, D_ARROW_KEYCODE].includes(e.key)) {
      if (focusIndex < refs.current.list.length - 1) {
        focusIndex++;
      } else {
        focusIndex = 0;
      }
      focusItem(focusIndex);
      // prevents page from scrolling
      e.preventDefault();
    } else if ([L_ARROW_KEYCODE, U_ARROW_KEYCODE].includes(e.key)) {
      if (focusIndex > 0) {
        focusIndex--;
      } else {
        focusIndex = refs.current.list.length - 1;
      }
      focusItem(focusIndex);
      // prevents page from scrolling
      e.preventDefault();
    }
  };

  const componentClassName = clsx(styles['dropdown-menu'], className);
  return (
    <DropdownMenuContext.Provider value={{ refs: refs }}>
      <div className={componentClassName} {...other}>
        <ul
          className={styles['dropdown-menu__list']}
          onKeyDown={onKeyDown}
          role="menu"
        >
          {children}
        </ul>
      </div>
    </DropdownMenuContext.Provider>
  );
};
