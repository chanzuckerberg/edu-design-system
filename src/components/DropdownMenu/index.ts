export { DropdownMenu as default, DropdownMenuContext } from './DropdownMenu';
