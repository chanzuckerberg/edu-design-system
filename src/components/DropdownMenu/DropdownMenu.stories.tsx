import { BADGE } from '@geometricpanda/storybook-addon-badges';
import type { StoryObj, Meta } from '@storybook/react';
import React from 'react';

import { DropdownMenu } from './DropdownMenu';
import { DropdownMenuItem } from '../DropdownMenuItem/DropdownMenuItem';
import { Icon } from '../Icon/Icon';

export default {
  title: 'Components/DropdownMenu',
  component: DropdownMenu,
  subcomponents: { DropdownMenuItem },
  parameters: {
    badges: [BADGE.DEPRECATED],
  },
} as Meta<Args>;

type Args = React.ComponentProps<typeof DropdownMenu>;

export const Default: StoryObj<Args> = {
  args: {
    children: (
      <>
        <DropdownMenuItem onClick={() => console.log('Item 1')}>
          Item 1
        </DropdownMenuItem>
        <DropdownMenuItem onClick={() => console.log('Item 2')}>
          Item 2
        </DropdownMenuItem>
        <DropdownMenuItem onClick={() => console.log('Item 3')} variant="lined">
          Item 3
        </DropdownMenuItem>
        <DropdownMenuItem href="#">Item 4</DropdownMenuItem>
      </>
    ),
  },
};

export const WithIcons: StoryObj<Args> = {
  args: {
    children: (
      <>
        <DropdownMenuItem>
          <Icon name="schedule" purpose="decorative" size="1.25rem" />
          Item 1
        </DropdownMenuItem>
        <DropdownMenuItem>
          <Icon name="schedule" purpose="decorative" size="1.25rem" />
          Item 2
        </DropdownMenuItem>
        <DropdownMenuItem>
          <Icon name="schedule" purpose="decorative" size="1.25rem" />
          Item 3
        </DropdownMenuItem>
        <DropdownMenuItem>
          <Icon name="schedule" purpose="decorative" size="1.25rem" />
          Item 4
        </DropdownMenuItem>
      </>
    ),
  },
};

export const WithDeleteButton: StoryObj<Args> = {
  args: {
    children: (
      <>
        <DropdownMenuItem>
          <Icon name="schedule" purpose="decorative" size="1.25rem" />
          Item 1
        </DropdownMenuItem>
        <DropdownMenuItem>
          <Icon name="schedule" purpose="decorative" size="1.25rem" />
          Item 2
        </DropdownMenuItem>
        <DropdownMenuItem>
          <Icon name="schedule" purpose="decorative" size="1.25rem" />
          Item 3
        </DropdownMenuItem>
        <DropdownMenuItem status="error">
          <Icon name="schedule" purpose="decorative" size="1.25rem" />
          Delete Item
        </DropdownMenuItem>
      </>
    ),
  },
};
