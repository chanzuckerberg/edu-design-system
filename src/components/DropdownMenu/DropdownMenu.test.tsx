import { generateSnapshots } from '@chanzuckerberg/story-utils';
import * as stories from './DropdownMenu.stories';

describe('<DropdownMenu />', () => {
  generateSnapshots(stories);
});
