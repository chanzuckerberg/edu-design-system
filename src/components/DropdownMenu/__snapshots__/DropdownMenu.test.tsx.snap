// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownMenu /> Default story renders snapshot 1`] = `
<div
  class="dropdown-menu"
>
  <ul
    class="dropdown-menu__list"
    role="menu"
  >
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        Item 1
      </button>
    </li>
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        Item 2
      </button>
    </li>
    <li
      class="dropdown-menu__item dropdown-menu__item--lined"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        Item 3
      </button>
    </li>
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <a
        class="dropdown-menu__link"
        href="#"
      >
        Item 4
      </a>
    </li>
  </ul>
</div>
`;

exports[`<DropdownMenu /> WithDeleteButton story renders snapshot 1`] = `
<div
  class="dropdown-menu"
>
  <ul
    class="dropdown-menu__list"
    role="menu"
  >
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.25rem"
          style="--icon-size: 1.25rem;"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#schedule"
          />
        </svg>
        Item 1
      </button>
    </li>
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.25rem"
          style="--icon-size: 1.25rem;"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#schedule"
          />
        </svg>
        Item 2
      </button>
    </li>
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.25rem"
          style="--icon-size: 1.25rem;"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#schedule"
          />
        </svg>
        Item 3
      </button>
    </li>
    <li
      class="dropdown-menu__item dropdown-menu__item--error"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.25rem"
          style="--icon-size: 1.25rem;"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#schedule"
          />
        </svg>
        Delete Item
      </button>
    </li>
  </ul>
</div>
`;

exports[`<DropdownMenu /> WithIcons story renders snapshot 1`] = `
<div
  class="dropdown-menu"
>
  <ul
    class="dropdown-menu__list"
    role="menu"
  >
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.25rem"
          style="--icon-size: 1.25rem;"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#schedule"
          />
        </svg>
        Item 1
      </button>
    </li>
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.25rem"
          style="--icon-size: 1.25rem;"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#schedule"
          />
        </svg>
        Item 2
      </button>
    </li>
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.25rem"
          style="--icon-size: 1.25rem;"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#schedule"
          />
        </svg>
        Item 3
      </button>
    </li>
    <li
      class="dropdown-menu__item"
      role="menuitem"
    >
      <button
        class="dropdown-menu__link"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.25rem"
          style="--icon-size: 1.25rem;"
          width="1.25rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#schedule"
          />
        </svg>
        Item 4
      </button>
    </li>
  </ul>
</div>
`;
